Crash while reading from xdata

This is very much reproducible: ~~read from `xdata` with negative `q0`~~. (see below)

```
panic: runtime error: slice bounds out of range [-2:]

goroutine 51 [running]:
panic(0x634ce0, 0xc000016960)
        /home/big/go/src/runtime/panic.go:706 +0x2c2 fp=0xc0001f7dd0 sp=0xc0001f7d40 pc=0x42ea42
runtime.goPanicSliceB(0xfffffffffffffffe, 0x2)
        /home/big/go/src/runtime/panic.go:87 +0xa3 fp=0xc0001f7e18 sp=0xc0001f7dd0 pc=0x42d543
main.xfidruneread(0xc0000c2000, 0xc00011a118, 0xcb0, 0xcb2, 0xc0001f7f30)
        /home/fhs/go/pkg/mod/github.com/rjkroege/edwood@v0.2.1-0.20190628123847-c80f7ecdf670/xfid.go:944 +0x397 fp=0xc0001f7ec8 sp=0xc0001f7e18 pc=0x5d43d7
main.xfidread(0xc0000c2000)
        /home/fhs/go/pkg/mod/github.com/rjkroege/edwood@v0.2.1-0.20190628123847-c80f7ecdf670/xfid.go:372 +0x9ba fp=0xc0001f7f88 sp=0xc0001f7ec8 pc=0x5d117a
main.xfidctl(0xc0000c2000, 0x69f500, 0xc000010120)
        /home/fhs/go/pkg/mod/github.com/rjkroege/edwood@v0.2.1-0.20190628123847-c80f7ecdf670/xfid.go:52 +0x9c fp=0xc0001f7fc8 sp=0xc0001f7f88 pc=0x5cf68c
runtime.goexit()
        /home/big/go/src/runtime/asm_amd64.s:1357 +0x1 fp=0xc0001f7fd0 sp=0xc0001f7fc8 pc=0x45bff1
created by main.xfidallocthread
        /home/fhs/go/pkg/mod/github.com/rjkroege/edwood@v0.2.1-0.20190628123847-c80f7ecdf670/acme.go:619 +0x1d4
```

--- a/xfid.go
+++ b/xfid.go
@@ -360,7 +360,7 @@ func xfidread(x *Xfid) {
 			x.respond(&fc, ErrAddrRange)
 			break
 		}
-		w.addr.q0 += xfidruneread(x, &w.body, (w.addr.q0), w.body.Nc())
+		w.addr.q0 += xfidruneread(x, &w.body, w.addr.q0, w.body.Nc())
 		w.addr.q1 = w.addr.q0
 
 	case QWxdata:
@@ -369,7 +369,7 @@ func xfidread(x *Xfid) {
 			x.respond(&fc, ErrAddrRange)
 			break
 		}
-		w.addr.q0 += xfidruneread(x, &w.body, (w.addr.q0), (w.addr.q1))
+		w.addr.q0 += xfidruneread(x, &w.body, w.addr.q0, w.addr.q1)
 
 	case QWtag:
 		xfidutfread(x, &w.tag, w.tag.Nc(), int(QWtag))
@@ -925,46 +925,37 @@ func xfidutfread(x *Xfid, t *Text, q1 int, qid int) {
 func xfidruneread(x *Xfid, t *Text, q0 int, q1 int) int {
 	// log.Println("xfidruneread", x)
 	// defer log.Println("done xfidruneread")
-	var (
-		fc plan9.Fcall
-		w  *Window
-	)
 
-	w = t.w
-	w.Commit(t)
+	t.w.Commit(t)
+
 	// Get Count runes, but that might be larger than Count bytes
 	nr := min(q1-q0, int(x.fcall.Count))
 	tmp := make([]rune, nr)
 	t.file.b.Read(q0, tmp)
 	buf := []byte(string(tmp))
-	// Now chop, and back up the end until we have a full rune
-	buf = buf[:nr]
-	i := nr - utf8.UTFMax
-	// Find a full rune to start in the last 4 bytes
-	for len(buf[i:]) > 0 {
-		ru, count := utf8.DecodeRune(buf[i:])
-		if ru == utf8.RuneError {
-			i += count
-		} else {
-			break
+
+	m := len(buf)
+	if len(buf) > int(x.fcall.Count) {
+		// copy whole runes only
+		m = 0
+		nr = 0
+		for m < len(buf) {
+			_, size := utf8.DecodeRune(buf[m:])
+			if m+size > int(x.fcall.Count) {
+				break
+			}
+			m += size
+			nr++
 		}
 	}
-	// add all further full runes
-	for len(buf[i:]) > 0 {
-		ru, count := utf8.DecodeRune(buf[i:])
-		if ru == utf8.RuneError {
-			break
-		} else {
-			i += count
-		}
+	buf = buf[:m]
+
+	fc := plan9.Fcall{
+		Count: uint32(len(buf)),
+		Data:  buf,
 	}
-
-	buf = buf[:i]
-
-	fc.Count = uint32(len(buf))
-	fc.Data = buf
 	x.respond(&fc, nil)
-	return len(string(buf))
+	return nr
 }
 
 func xfideventread(x *Xfid, w *Window) {

--- a/xfid_test.go
+++ b/xfid_test.go
@@ -63,3 +63,66 @@ func TestFullrunewrite(t *testing.T) {
 		}
 	}
 }
+
+type mockResponder struct {
+	fcall *plan9.Fcall
+	err   error
+}
+
+func (mr *mockResponder) respond(x *Xfid, t *plan9.Fcall, err error) *Xfid {
+	mr.fcall = t
+	mr.err = err
+	return x
+}
+
+func (mr *mockResponder) msize() int { return 8192 }
+
+func TestXfidruneread(t *testing.T) {
+	tt := []struct {
+		body   []rune // window body
+		q0, q1 int
+		count  uint32 // input fcall count
+		nr     int    // return value (number of runes)
+		data   []byte // output fcall data
+	}{
+		{[]rune("abcde"), 0, 5, 100, 5, []byte("abcde")},
+		{[]rune("abcde"), 1, 5, 100, 4, []byte("bcde")},
+		{[]rune("abcde"), 2, 5, 100, 3, []byte("cde")},
+		{[]rune("abcde"), 3, 5, 100, 2, []byte("de")},
+		{[]rune("abcde"), 4, 5, 100, 1, []byte("e")},
+		{[]rune("abcde"), 5, 5, 100, 0, []byte("")},
+		{[]rune("αβξδε"), 0, 5, 100, 5, []byte("αβξδε")},
+		{[]rune("αβξδε"), 1, 5, 100, 4, []byte("βξδε")},
+		{[]rune("αβξδε"), 2, 5, 100, 3, []byte("ξδε")},
+		{[]rune("αβξδε"), 3, 5, 100, 2, []byte("δε")},
+		{[]rune("αβξδε"), 4, 5, 100, 1, []byte("ε")},
+		{[]rune("αβξδε"), 0, 5, 8, 4, []byte("αβξδ")},
+		{[]rune("αβξδε"), 0, 5, 5, 2, []byte("αβ")},
+		{[]rune("αβξδε"), 0, 5, 0, 0, []byte("")},
+	}
+
+	for _, tc := range tt {
+		mr := new(mockResponder)
+		x := &Xfid{
+			fcall: plan9.Fcall{
+				Count: tc.count,
+			},
+			fs: mr,
+		}
+		w := NewWindow().initHeadless(nil)
+		w.body.file.b = Buffer(tc.body)
+		nr := xfidruneread(x, &w.body, tc.q0, tc.q1)
+		if got, want := nr, tc.nr; got != want {
+			t.Errorf("read %v runes from %q (q0=%v, q1=%v); should read %v runes",
+				got, tc.body, tc.q0, tc.q1, want)
+		}
+		if got, want := mr.fcall.Count, uint32(len(tc.data)); got != want {
+			t.Errorf("read %v bytes from %q (q0=%v, q1=%v); want %v",
+				got, tc.body, tc.q0, tc.q1, want)
+		}
+		if got, want := mr.fcall.Data, tc.data; !bytes.Equal(got, want) {
+			t.Errorf("read %q from %q (q0=%v, q1=%v); want %q\n",
+				got, tc.body, tc.q0, tc.q1, want)
+		}
+	}
+}

